### PR TITLE
Verifying multi-entity detectors

### PR DESCRIPTION
**Pull request by kaituo**
*Issue #, if available:*

*Description of changes:*

This PR adds categorical fields' number and type check. We only support one categorical field, and the categorical field can only be of type keyword and ip.  We also limit the max multi-entity detectors to 10.

Testing done:
1. added unit tests
2. did manual testing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/IndexAnomalyDetectorActionHandler.java [hunk 1]
```diff
@@ -178,32 +196,130 @@ private void onGetAnomalyDetectorResponse(GetResponse response) throws IOExcepti
...
+            String errorMsg = EXCEEDED_MAX_MULTI_ENTITY_DETECTORS_PREFIX_MSG + maxMultiEntityAnomalyDetectors;
+            logger.error(errorMsg);
+            onFailure(new IllegalArgumentException(errorMsg));
+        } else {
+            validateCategoricalField(null);
+        }
+    }
+
+    /**
+     * Precondition: anomalyDetector.getCategoryField() != null.
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/IndexAnomalyDetectorActionHandler.java:255]: What does this comment mean?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 2]
```diff
@@ -105,7 +107,62 @@
...
+        if (indices == null || indices.isEmpty()) {
+            throw new IllegalArgumentException("Indices should be set");
+        }
+        if (detectionInterval == null) {
+            throw new IllegalArgumentException("Detection interval should be set");
+        }
+        if (shingleSize != null && shingleSize < 1) {
+            throw new IllegalArgumentException("Shingle size must be a positive integer");
+        }
+        if (categoryField != null && categoryField.size() > 1) {
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:144]: 1 is magic number, how about we create CATEGORY_FIELD_LIMIT instant and use it at all other places. like IndexAnomalyDetectorActionHandler.java line 269: 
`if (categoryField.size() != 1) {`

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 3]
```diff
@@ -87,6 +88,7 @@
     private final Map<String, Object> uiMetadata;
     private final Integer schemaVersion;
     private final Instant lastUpdateTime;
+    private final List<String> categoryField;
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:91]: If we plan to support multiple category fields, should we name this as "categoryFields" ?

**Review by saratvemulapalli — COMMENTED** [review]
Overall the changes look good to me.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (saratvemulapalli) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:144]: +1 I like the simpler check.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 3 above)
Comment (saratvemulapalli) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:91]: I see we just use 1 category field today, do we see possible use cases of multiple category fields ?

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:91]: yes, changed

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java]: yes, maybe in the future

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java]: changed

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/IndexAnomalyDetectorActionHandler.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/IndexAnomalyDetectorActionHandler.java]: not true anymore.  Removed.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java [hunk 4]
```diff
@@ -22,6 +22,9 @@
     // index name for anomaly checkpoint of each model. One model one document.
     public static final String CHECKPOINT_INDEX_NAME = ".opendistro-anomaly-checkpoints";
 
+    // The alias of the index in which to write single-entity AD result history
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java:25]: Do we plan to write HC detector's result into another index?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 5]
```diff
@@ -72,6 +75,7 @@
     private static final String SHINGLE_SIZE_FIELD = "shingle_size";
     private static final String LAST_UPDATE_TIME_FIELD = "last_update_time";
     public static final String UI_METADATA_FIELD = "ui_metadata";
+    public static final String CATEGORY_FIELD = "category_field";
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:78]: Should we change to "category_fields" ?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java (hunk 4 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java:25]: not in the plan.  The code would have to change a lot for that to happen.  e.g., depends on there is categorical field or not, job scheduler will save to different places.  Right now, job scheduler does not have access to AnomalyDetector object.

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java:25]: Got it, how about remove the "single-entity" from the comments?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 5 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:78]: Can we do it later?  Both Yizhe and I need to change for this.

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:78]: Sure, I'm ok

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java (hunk 4 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonName.java:25]: removed.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 5 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:78]: Thanks so much for your consideration.

**Review by ylwu-amzn — APPROVED** [review]
LGTM. Thanks for the change

**Review by saratvemulapalli — APPROVED** [review]
Changes look good to me.

**Pull request merged**